### PR TITLE
feat: added subprocess test client

### LINKED ISSUE #3654: Enhancement: Provide test client that can handle infinite SSE generators
### Summary

Follow-up from https://github.com/orgs/litestar-org/discussions/3547

The litestart test clients directly call the ASGI app instance. When calling stream endpoints like SSEs, they will always try to consume the whole stream. If the endpoint serves an infinite generator, the test client is stuck. This does not happen for the test client websocket implementation is because it "cheats" by not implementing the actual websocket protocol but using an internal queue to transport the messages.

The solution for this edge case is to run the app in a subprocess (e.g. via uvicorn) and then use an httpx client to connect to the web app.

### Basic Example

_No response_

### Drawbacks and Impact

_No response_

### Unresolved questions

_No response_

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3654">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3654/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by aranvir**
## Description

This introduces new helper functions to create sync and async test clients that can be used to test SSE endpoints with infinite generators as outlined in #3654 and https://github.com/orgs/litestar-org/discussions/3547.

The litestar cli is run in a subprocess to start a web app instance with a random port. The helper function then sets up a basic sync or async client.

The test code and doc example show how this can be used to test an SSE endpoint with an infinite generator.

Overview:
* Adds `subprocess_async_client` and `subprocess_sync_client`
* Tests both clients
* Adds general documentation. Adds docs example for `subprocess_async_client`

## Closes
Closes #3654

### CODE REVIEW

**Comment by aranvir** [comment]
Tests are failing because the redis fixture does not seem to work correctly. Tbh, added that as a best guess, so would need some help with fixing this. Alternatively, I just replace the test with a regular "does it connect and return the right thing" kind of test and ditch the SSE testing.

No idea why doc generation is failing.

**Comment by JacobCoffee** [comment]
> Tests are failing because the redis fixture does not seem to work correctly. Tbh, added that as a best guess, so would need some help with fixing this. Alternatively, I just replace the test with a regular "does it connect and return the right thing" kind of test and ditch the SSE testing.

cc: @litestar-org/maintainers

File: docs/usage/testing.rst [hunk 1]
```diff
@@ -279,6 +279,19 @@ But also this:
             assert response.text == "healthy"
 
 
+Creating an external test app
+-------------------
```
Comment (provinzkraut) [inline docs/usage/testing.rst:283]:
```suggestion
Creating an external test app
------------------------------
```

File: docs/usage/testing.rst [hunk 2]
```diff
@@ -279,6 +279,19 @@ But also this:
...
+
+The test clients make use of the capability to directly load an ASGI app into an httpx Client without having to run an actual server like uvicorn. For most test cases, this is sufficient but there are some situations where this will not work. For example, when using server-sent events with an infinite generator, it will lock up the test client, since it tries to first exhaust the generator and then return to the test function.
+
+Litestar offers two helper functions called :func:`subprocess_sync_client <litestar.testing.client.subprocess_client.subprocess_sync_client>` and :func:`subprocess_async_client <litestar.testing.client.subprocess_client.subprocess_async_client>` that will launch a litestar instance with uvicorn in a subprocess and set up an httpx client for running tests. You can either load your actual app file or create subsets from it as you would with the regular test client setup. An example is shown below.
```
Comment (provinzkraut) [inline docs/usage/testing.rst:287]:
```suggestion
Litestar offers two helper functions, :func:`~litestar.testing.client.subprocess_client.subprocess_sync_client` and :func:`~litestar.testing.client.subprocess_client.subprocess_async_client` that will launch a Litestar instance with in a subprocess and set up an httpx client for running tests. You can either load your actual app file or create subsets from it as you would with the regular test client setup. An example is shown below.
```

File: docs/usage/testing.rst [hunk 3]
```diff
@@ -279,6 +279,19 @@ But also this:
             assert response.text == "healthy"
 
 
+Creating an external test app
+-------------------
+
+The test clients make use of the capability to directly load an ASGI app into an httpx Client without having to run an actual server like uvicorn. For most test cases, this is sufficient but there are some situations where this will not work. For example, when using server-sent events with an infinite generator, it will lock up the test client, since it tries to first exhaust the generator and then return to the test function.
```
Comment (provinzkraut) [inline docs/usage/testing.rst:285]:
```suggestion
The test clients make use of the capability to directly load an ASGI app into an httpx Client without having to run an actual server like uvicorn. For most test cases, this is sufficient but there are some situations where this will not work. For example, when using server-sent events with an infinite generator, due to the way HTTPX (which the Litestar test clients are built upon) works internally with ASGI applications, it will lock up the test client, since it tries to first exhaust the generator and then return to the test function.
```

**Review by provinzkraut — APPROVED** [review]

**Review by cofin — APPROVED** [review]
Simplified the test case a bit

**Comment by cofin** [comment]
@all-contributors add @aranvir for docs, code, tests
